vumi_message_store/memory_backend_manager.py

```diff
@@ -165,18 +165,28 @@ def store_object(self, fake_object):
             "content_type": fake_object._current_data["content_type"],
             "data": json.dumps(fake_object._current_data["data"]),
         }
-        indexes = {}
+
+        bucket["indexes"] = (
+            self._rebuild_bucket_indexes(fake_object, bucket["indexes"]))
+
+    def _rebuild_bucket_indexes(self, fake_object, bucket_indexes):
+        """
+        Clears any existing indexes for the object and adds the new indexes.
+        """
+        new_bucket_indexes = {}
+        for index_name, index_values in bucket_indexes.iteritems():
+            new_index_values = list(filter(
+                lambda (_, key): key != fake_object.key, index_values))
```

Comment:
I have a mild preference for the listcomp version (mostly because it's not immediately obvious where them lambda ends) but I'm happy to leave this as-is.